File: database-only/src/com/performance/util/DatabaseManager.java
package com.performance.util;

import com.performance.model.*;
import java.sql.*;
import java.util.*;

/**
 * Database Manager for MySQL database operations.
 */
public class DatabaseManager {
    // MySQL connection URL - update host, port, and database name as needed
    private static final String DB_URL = "jdbc:mysql://localhost:3306/performance";
    private static final String DB_USER = "root"; // Update with your MySQL username
    private static final String DB_PASSWORD = ""; // Update with your MySQL password

    private static DatabaseManager instance;
    private Connection connection;

    private DatabaseManager() {
        try {
            Class.forName("com.mysql.cj.jdbc.Driver");
            connection = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
            createTables();
        } catch (Exception e) {
            System.err.println("Database connection failed: " + e.getMessage());
            e.printStackTrace();
        }
    }

    public static synchronized DatabaseManager getInstance() {
        if (instance == null) {
            instance = new DatabaseManager();
        }
        return instance;
    }

    private void createTables() throws SQLException {
        Statement stmt = connection.createStatement();

        // Users table
        stmt.execute("""
                    CREATE TABLE IF NOT EXISTS users (
                        id INT PRIMARY KEY AUTO_INCREMENT,
                        username VARCHAR(100) UNIQUE NOT NULL,
                        password VARCHAR(255) NOT NULL,
                        name VARCHAR(255) NOT NULL,
                        role VARCHAR(50) DEFAULT 'EMPLOYEE',
                        status VARCHAR(50) DEFAULT 'PENDING',
                        tier_level INT DEFAULT 1,
                        manager_id INT,
                        FOREIGN KEY (manager_id) REFERENCES users(id)
                    )
                """);

        // Reviews table
        stmt.execute("""
                    CREATE TABLE IF NOT EXISTS reviews (
                        id INT PRIMARY KEY AUTO_INCREMENT,
                        employee_id INT NOT NULL,
                        reviewer_id INT,
                        punctuality INT DEFAULT 0,
                        teamwork INT DEFAULT 0,
                        productivity INT DEFAULT 0,
                        leadership INT DEFAULT 0,
                        vision INT DEFAULT 0,
                        managerial_skills INT DEFAULT 0,
                        comments TEXT,
                        status VARCHAR(50) DEFAULT 'SUBMITTED',
                        created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                        FOREIGN KEY (employee_id) REFERENCES users(id),
                        FOREIGN KEY (reviewer_id) REFERENCES users(id)
                    )
                """);

        // Goals table
        stmt.execute("""
                    CREATE TABLE IF NOT EXISTS goals (
                        id INT PRIMARY KEY AUTO_INCREMENT,
                        employee_id INT NOT NULL,
                        creator_id INT,
                        description TEXT NOT NULL,
                        status VARCHAR(50) DEFAULT 'ACTIVE',
                        created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                        FOREIGN KEY (employee_id) REFERENCES users(id),
                        FOREIGN KEY (creator_id) REFERENCES users(id)
                    )
                """);

        // Training recommendations table
        stmt.execute("""
                    CREATE TABLE IF NOT EXISTS training_recommendations (
                        id INT PRIMARY KEY AUTO_INCREMENT,
                        recipient_id INT NOT NULL,
                        recommender_id INT,
                        course_name VARCHAR(255) NOT NULL,
                        reason TEXT,
                        created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                        FOREIGN KEY (recipient_id) REFERENCES users(id),
                        FOREIGN KEY (recommender_id) REFERENCES users(id)
                    )
                """);

        // Courses table
        stmt.execute("""
                    CREATE TABLE IF NOT EXISTS courses (
                        id INT PRIMARY KEY AUTO_INCREMENT,
                        name VARCHAR(255) UNIQUE NOT NULL
                    )
                """);

        stmt.close();
    }

    // === USER OPERATIONS ===

    public void saveUser(User user) {
        try {
            // Check if exists
            PreparedStatement check = connection.prepareStatement(
                    "SELECT id FROM users WHERE username = ?");
            check.setString(1, user.getUsername());
            ResultSet rs = check.executeQuery();

            if (rs.next()) {
                // Update existing
                int id = rs.getInt("id");
                PreparedStatement update = connection.prepareStatement("""
                            UPDATE users SET password=?, name=?, role=?, status=?, tier_level=?, manager_id=?
                            WHERE id=?
                        """);
                update.setString(1, user.getPassword());
                update.setString(2, user.getName());
                update.setString(3, user.getRole().toString());
                update.setString(4, user.getStatus().toString());
                update.setInt(5, user.getTierLevel());
                if (user.getManager() != null) {
                    update.setInt(6, getUserId(user.getManager().getUsername()));
                } else {
                    update.setNull(6, Types.INTEGER);
                }
                update.setInt(7, id);
                update.executeUpdate();
            } else {
                // Insert new
                PreparedStatement insert = connection.prepareStatement("""
                            INSERT INTO users (username, password, name, role, status, tier_level, manager_id)
                            VALUES (?, ?, ?, ?, ?, ?, ?)
                        """);
                insert.setString(1, user.getUsername());
                insert.setString(2, user.getPassword());
                insert.setString(3, user.getName());
                insert.setString(4, user.getRole().toString());
                insert.setString(5, user.getStatus().toString());
                insert.setInt(6, user.getTierLevel());
                if (user.getManager() != null) {
                    insert.setInt(7, getUserId(user.getManager().getUsername()));
                } else {
                    insert.setNull(7, Types.INTEGER);
                }
                insert.executeUpdate();
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public int getUserId(String username) {
        try {
            PreparedStatement stmt = connection.prepareStatement(
                    "SELECT id FROM users WHERE username = ?");
            stmt.setString(1, username);
            ResultSet rs = stmt.executeQuery();
            if (rs.next())
                return rs.getInt("id");
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public List<User> getAllUsers() {
        List<User> users = new ArrayList<>();
        Map<Integer, User> userMap = new HashMap<>();
        Map<Integer, Integer> managerMap = new HashMap<>();

        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM users");

            while (rs.next()) {
                int id = rs.getInt("id");
                String username = rs.getString("username");
                String password = rs.getString("password");
                String name = rs.getString("name");
                String role = rs.getString("role");
                String status = rs.getString("status");
                int tierLevel = rs.getInt("tier_level");
                int managerId = rs.getInt("manager_id");

                User user;
                if ("ADMIN".equals(role)) {
                    user = new Admin(username, password, name);
                } else {
                    user = new Employee(username, password, name);
                    user.setRole(UserRole.valueOf(role));
                }
                user.setStatus(UserStatus.valueOf(status));
                user.setTierLevel(tierLevel);

                userMap.put(id, user);
                if (managerId > 0) {
                    managerMap.put(id, managerId);
                }
                users.add(user);
            }

            // Set managers
            for (Map.Entry<Integer, Integer> entry : managerMap.entrySet()) {
                User user = userMap.get(entry.getKey());
                User manager = userMap.get(entry.getValue());
                if (user != null && manager != null) {
                    user.setManager(manager);
                    manager.addDirectReport(user);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return users;
    }

    public void deleteUser(String username) {
        try {
            PreparedStatement stmt = connection.prepareStatement(
                    "DELETE FROM users WHERE username = ?");
            stmt.setString(1, username);
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // === REVIEW OPERATIONS ===

    public void saveReview(Review review) {
        try {
            PreparedStatement stmt = connection.prepareStatement("""
                        INSERT INTO reviews (employee_id, reviewer_id, punctuality, teamwork, productivity,
                            leadership, vision, managerial_skills, comments, status)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """);
            stmt.setInt(1, getUserId(review.getEmployee().getUsername()));
            stmt.setInt(2, review.getReviewer() != null ? getUserId(review.getReviewer().getUsername()) : 0);
            stmt.setInt(3, review.getPunctualityScore());
            stmt.setInt(4, review.getTeamworkScore());
            stmt.setInt(5, review.getProductivityScore());
            stmt.setInt(6, review.getLeadershipScore());
            stmt.setInt(7, review.getVisionScore());
            stmt.setInt(8, review.getManagerialSkillsScore());
            stmt.setString(9, review.getComments());
            stmt.setString(10, review.getStatus());
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public List<Review> getAllReviews(Map<Integer, User> userMap) {
        List<Review> reviews = new ArrayList<>();
        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM reviews");

            while (rs.next()) {
                User employee = userMap.get(rs.getInt("employee_id"));
                User reviewer = userMap.get(rs.getInt("reviewer_id"));
                if (employee != null) {
                    Review review = new Review(employee, reviewer);
                    review.setPunctualityScore(rs.getInt("punctuality"));
                    review.setTeamworkScore(rs.getInt("teamwork"));
                    review.setProductivityScore(rs.getInt("productivity"));
                    review.setLeadershipScore(rs.getInt("leadership"));
                    review.setVisionScore(rs.getInt("vision"));
                    review.setManagerialSkillsScore(rs.getInt("managerial_skills"));
                    review.setComments(rs.getString("comments"));
                    review.setStatus(rs.getString("status"));
                    reviews.add(review);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return reviews;
    }

    // === GOAL OPERATIONS ===

    public void saveGoal(Goal goal) {
        try {
            PreparedStatement stmt = connection.prepareStatement("""
                        INSERT INTO goals (employee_id, creator_id, description, status)
                        VALUES (?, ?, ?, ?)
                    """);
            stmt.setInt(1, getUserId(goal.getEmployee().getUsername()));
            stmt.setInt(2, goal.getCreator() != null ? getUserId(goal.getCreator().getUsername()) : 0);
            stmt.setString(3, goal.getDescription());
            stmt.setString(4, goal.getStatus());
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public List<Goal> getAllGoals(Map<Integer, User> userMap) {
        List<Goal> goals = new ArrayList<>();
        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM goals");

            while (rs.next()) {
                User employee = userMap.get(rs.getInt("employee_id"));
                User creator = userMap.get(rs.getInt("creator_id"));
                if (employee != null) {
                    Goal goal = new Goal(employee, rs.getString("description"), creator);
                    goal.setStatus(rs.getString("status"));
                    goals.add(goal);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return goals;
    }

    public void deleteGoal(Goal goal) {
        try {
            PreparedStatement stmt = connection.prepareStatement(
                    "DELETE FROM goals WHERE employee_id = ? AND description = ?");
            stmt.setInt(1, getUserId(goal.getEmployee().getUsername()));
            stmt.setString(2, goal.getDescription());
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // === COURSE OPERATIONS ===

    public void saveCourse(String courseName) {
        try {
            PreparedStatement stmt = connection.prepareStatement(
                    "INSERT IGNORE INTO courses (name) VALUES (?)");
            stmt.setString(1, courseName);
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public List<String> getAllCourses() {
        List<String> courses = new ArrayList<>();
        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT name FROM courses");
            while (rs.next()) {
                courses.add(rs.getString("name"));
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return courses;
    }

    public void deleteCourse(String name) {
        try {
            PreparedStatement stmt = connection.prepareStatement(
                    "DELETE FROM courses WHERE name = ?");
            stmt.setString(1, name);
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    // === TRAINING RECOMMENDATIONS ===

    public void saveTrainingRecommendation(TrainingRecommendation tr) {
        try {
            PreparedStatement stmt = connection.prepareStatement("""
                        INSERT INTO training_recommendations (recipient_id, recommender_id, course_name)
                        VALUES (?, ?, ?)
                    """);
            stmt.setInt(1, getUserId(tr.getRecipient().getUsername()));
            stmt.setInt(2, tr.getRecommender() != null ? getUserId(tr.getRecommender().getUsername()) : 0);
            stmt.setString(3, tr.getCourseName());
            stmt.executeUpdate();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    public List<TrainingRecommendation> getAllTrainingRecommendations(Map<Integer, User> userMap) {
        List<TrainingRecommendation> recommendations = new ArrayList<>();
        try {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM training_recommendations");

            while (rs.next()) {
                User recipient = userMap.get(rs.getInt("recipient_id"));
                User recommender = userMap.get(rs.getInt("recommender_id"));
                if (recipient != null) {
                    TrainingRecommendation tr = new TrainingRecommendation(
                            recommender, recipient, rs.getString("course_name"));
                    recommendations.add(tr);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return recommendations;
    }

    public void close() {
        try {
            if (connection != null)
                connection.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }
}

File: project-dotnet/PerformanceReview/Utilities/DatabaseManager.cs
using Microsoft.Data.Sqlite;
using PerformanceReview.Models;
using System;
using System.Collections.Generic;

namespace PerformanceReview.Utilities
{
    public class DatabaseManager
    {
        private static DatabaseManager? _instance;
        private SqliteConnection _connection;
        private const string DB_URL = "Data Source=performance.db";

        private DatabaseManager()
        {
            _connection = new SqliteConnection(DB_URL);
            _connection.Open();
            CreateTables();
        }

        public static DatabaseManager GetInstance()
        {
            if (_instance == null) _instance = new DatabaseManager();
            return _instance;
        }

        private void CreateTables()
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = @"
                CREATE TABLE IF NOT EXISTS users (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    username TEXT UNIQUE NOT NULL,
                    password TEXT NOT NULL,
                    name TEXT NOT NULL,
                    role TEXT DEFAULT 'EMPLOYEE',
                    status TEXT DEFAULT 'PENDING',
                    tier_level INTEGER DEFAULT 1,
                    manager_id INTEGER,
                    FOREIGN KEY (manager_id) REFERENCES users(id)
                );
                CREATE TABLE IF NOT EXISTS reviews (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    employee_id INTEGER NOT NULL,
                    reviewer_id INTEGER,
                    punctuality INTEGER DEFAULT 0,
                    teamwork INTEGER DEFAULT 0,
                    productivity INTEGER DEFAULT 0,
                    leadership INTEGER DEFAULT 0,
                    vision INTEGER DEFAULT 0,
                    managerial_skills INTEGER DEFAULT 0,
                    comments TEXT,
                    status TEXT DEFAULT 'SUBMITTED',
                    created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                    FOREIGN KEY (employee_id) REFERENCES users(id),
                    FOREIGN KEY (reviewer_id) REFERENCES users(id)
                );
                CREATE TABLE IF NOT EXISTS goals (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    employee_id INTEGER NOT NULL,
                    creator_id INTEGER,
                    description TEXT NOT NULL,
                    status TEXT DEFAULT 'ACTIVE',
                    created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                    FOREIGN KEY (employee_id) REFERENCES users(id),
                    FOREIGN KEY (creator_id) REFERENCES users(id)
                );
                CREATE TABLE IF NOT EXISTS training_recommendations (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    recipient_id INTEGER NOT NULL,
                    recommender_id INTEGER,
                    course_name TEXT NOT NULL,
                    created_date DATETIME DEFAULT CURRENT_TIMESTAMP,
                    FOREIGN KEY (recipient_id) REFERENCES users(id),
                    FOREIGN KEY (recommender_id) REFERENCES users(id)
                );
                CREATE TABLE IF NOT EXISTS courses (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    name TEXT UNIQUE NOT NULL
                );";
            cmd.ExecuteNonQuery();
        }

        public void SaveUser(User user)
        {
            try
            {
                using var check = _connection.CreateCommand();
                check.CommandText = "SELECT id FROM users WHERE username = @username";
                check.Parameters.AddWithValue("@username", user.Username);
                var result = check.ExecuteScalar();

                if (result != null)
                {
                    using var update = _connection.CreateCommand();
                    update.CommandText = @"UPDATE users SET password=@p, name=@n, role=@r, status=@s, 
                        tier_level=@t, manager_id=@m WHERE id=@id";
                    update.Parameters.AddWithValue("@p", user.Password);
                    update.Parameters.AddWithValue("@n", user.Name);
                    update.Parameters.AddWithValue("@r", user.Role.ToString());
                    update.Parameters.AddWithValue("@s", user.Status.ToString());
                    update.Parameters.AddWithValue("@t", user.TierLevel);
                    update.Parameters.AddWithValue("@m", user.Manager != null ? (object)GetUserId(user.Manager.Username) : DBNull.Value);
                    update.Parameters.AddWithValue("@id", (long)result);
                    update.ExecuteNonQuery();
                }
                else
                {
                    using var insert = _connection.CreateCommand();
                    insert.CommandText = @"INSERT INTO users (username, password, name, role, status, tier_level, manager_id) 
                        VALUES (@u, @p, @n, @r, @s, @t, @m)";
                    insert.Parameters.AddWithValue("@u", user.Username);
                    insert.Parameters.AddWithValue("@p", user.Password);
                    insert.Parameters.AddWithValue("@n", user.Name);
                    insert.Parameters.AddWithValue("@r", user.Role.ToString());
                    insert.Parameters.AddWithValue("@s", user.Status.ToString());
                    insert.Parameters.AddWithValue("@t", user.TierLevel);
                    insert.Parameters.AddWithValue("@m", user.Manager != null ? (object)GetUserId(user.Manager.Username) : DBNull.Value);
                    insert.ExecuteNonQuery();
                }
            }
            catch (Exception e) { Console.Error.WriteLine(e.Message); }
        }

        public int GetUserId(string username)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT id FROM users WHERE username = @u";
            cmd.Parameters.AddWithValue("@u", username);
            var result = cmd.ExecuteScalar();
            return result != null ? Convert.ToInt32(result) : -1;
        }

        public List<User> GetAllUsers()
        {
            var users = new List<User>();
            var userMap = new Dictionary<int, User>();
            var managerMap = new Dictionary<int, int>();

            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT * FROM users";
            using var reader = cmd.ExecuteReader();

            while (reader.Read())
            {
                int id = reader.GetInt32(0);
                string username = reader.GetString(1);
                string password = reader.GetString(2);
                string name = reader.GetString(3);
                string role = reader.GetString(4);
                string status = reader.GetString(5);
                int tierLevel = reader.GetInt32(6);
                int managerId = reader.IsDBNull(7) ? 0 : reader.GetInt32(7);

                User user;
                if (role == "ADMIN")
                    user = new Admin(username, password, name);
                else
                {
                    user = new Employee(username, password, name);
                    user.Role = Enum.Parse<UserRole>(role);
                }
                user.Status = Enum.Parse<UserStatus>(status);
                user.TierLevel = tierLevel;

                userMap[id] = user;
                if (managerId > 0) managerMap[id] = managerId;
                users.Add(user);
            }

            foreach (var entry in managerMap)
            {
                if (userMap.TryGetValue(entry.Key, out var user) && userMap.TryGetValue(entry.Value, out var manager))
                {
                    user.Manager = manager;
                    manager.AddDirectReport(user);
                }
            }

            return users;
        }

        public void DeleteUser(string username)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "DELETE FROM users WHERE username = @u";
            cmd.Parameters.AddWithValue("@u", username);
            cmd.ExecuteNonQuery();
        }

        public void SaveReview(Review review)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = @"INSERT INTO reviews (employee_id, reviewer_id, punctuality, teamwork, productivity,
                leadership, vision, managerial_skills, comments, status) VALUES (@e, @r, @p, @t, @pr, @l, @v, @ms, @c, @s)";
            cmd.Parameters.AddWithValue("@e", GetUserId(review.Employee.Username));
            cmd.Parameters.AddWithValue("@r", review.Reviewer != null ? (object)GetUserId(review.Reviewer.Username) : 0);
            cmd.Parameters.AddWithValue("@p", review.PunctualityScore);
            cmd.Parameters.AddWithValue("@t", review.TeamworkScore);
            cmd.Parameters.AddWithValue("@pr", review.ProductivityScore);
            cmd.Parameters.AddWithValue("@l", review.LeadershipScore);
            cmd.Parameters.AddWithValue("@v", review.VisionScore);
            cmd.Parameters.AddWithValue("@ms", review.ManagerialSkillsScore);
            cmd.Parameters.AddWithValue("@c", (object?)review.Comments ?? DBNull.Value);
            cmd.Parameters.AddWithValue("@s", review.Status);
            cmd.ExecuteNonQuery();
        }

        public void SaveGoal(Goal goal)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "INSERT INTO goals (employee_id, creator_id, description, status) VALUES (@e, @c, @d, @s)";
            cmd.Parameters.AddWithValue("@e", GetUserId(goal.Employee.Username));
            cmd.Parameters.AddWithValue("@c", goal.Creator != null ? (object)GetUserId(goal.Creator.Username) : 0);
            cmd.Parameters.AddWithValue("@d", goal.Description);
            cmd.Parameters.AddWithValue("@s", goal.Status);
            cmd.ExecuteNonQuery();
        }

        public void SaveCourse(string courseName)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "INSERT OR IGNORE INTO courses (name) VALUES (@n)";
            cmd.Parameters.AddWithValue("@n", courseName);
            cmd.ExecuteNonQuery();
        }

        public List<string> GetAllCourses()
        {
            var courses = new List<string>();
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT name FROM courses";
            using var reader = cmd.ExecuteReader();
            while (reader.Read()) courses.Add(reader.GetString(0));
            return courses;
        }

        public void SaveTrainingRecommendation(TrainingRecommendation tr)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "INSERT INTO training_recommendations (recipient_id, recommender_id, course_name) VALUES (@r, @rc, @c)";
            cmd.Parameters.AddWithValue("@r", GetUserId(tr.Recipient.Username));
            cmd.Parameters.AddWithValue("@rc", tr.Recommender != null ? (object)GetUserId(tr.Recommender.Username) : 0);
            cmd.Parameters.AddWithValue("@c", tr.CourseName);
            cmd.ExecuteNonQuery();
        }

        public void Close()
        {
            _connection?.Close();
        }
    }
}
